In [1]:
# import every tool you might need for the project
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import requests
import re
import tweepy
import json
import os
from io import BytesIO
from PIL import Image

## Gather

In [2]:
df = pd.read_csv("twitter-archive.csv")
df.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [3]:
# create a directory first then save images in the directory
folder_name = 'image_predictions'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

with open(os.path.join(folder_name,
                      url.split('/')[-1]), mode='wb') as file:
    file.write(response.content)

In [4]:
#open and asign the downloaded tsv file to images
images = pd.read_csv('image_predictions/image-predictions.tsv', delimiter='\t')

In [5]:
# my personal API keys, secrets, and tokens which i m going to hide later before submission
consumer_key = 'LfZKUzhktxY4Jt8MMxCcjUsAN'
consumer_secret = 'OssH2QGJzQPxpeBs5qKCpOL5BujnfBmajPNubyS7XzGOAu1QHB'
access_token = '1077719655463047168-PNvDXrMmVE7tCLIAvFOYc8Cft90krg'
access_secret = 'UGT3wdrF18DG96ivSLc0b7cF6WIIl0htbG2Cuuo0MtOMh'

In [6]:
# tweepy authentication process

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth_handler=auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [7]:
# for step 3, get tweets info using list of tweet ids and record the failed id's then read into a .txt file

tweet_ids = list(df.tweet_id)
tweet_data = {}

for tweet in tweet_ids:
    try:
        tweet_status = api.get_status(tweet, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)
        tweet_data[str(tweet)] = tweet_status._json
    except: 
        print("Error: " + str(tweet))

Error: 888202515573088257
Error: 873697596434513921
Error: 872668790621863937
Error: 869988702071779329
Error: 866816280283807744
Error: 861769973181624320
Error: 845459076796616705
Error: 842892208864923648
Error: 837012587749474308
Error: 827228250799742977
Error: 812747805718642688
Error: 802247111496568832
Error: 775096608509886464
Error: 770743923962707968


Rate limit reached. Sleeping for: 642


Error: 754011816964026368
Error: 680055455951884288


Rate limit reached. Sleeping for: 664


In [8]:
with open('tweet_json.txt', 'w', encoding='utf-8') as outfile:  
    json.dump(tweet_data, outfile, sort_keys = True, indent=4, ensure_ascii = False)

In [9]:
# read the json file with index set as the orientation
df_tweet = pd.read_json('tweet_json.txt', orient='index')

### Assess

In [10]:
df.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [11]:
df_tweet.head(2)

contributors  coordinates          created_at  \
1991-02-08 13:48:08.022790149           NaN          NaN 2015-11-15 22:32:08   
1991-02-08 16:08:05.002620928           NaN          NaN 2015-11-15 23:05:30   

                                                                        entities  \
1991-02-08 13:48:08.022790149  {'hashtags': [], 'media': [{'display_url': 'pi...   
1991-02-08 16:08:05.002620928  {'hashtags': [], 'media': [{'display_url': 'pi...   

                                                               extended_entities  \
1991-02-08 13:48:08.022790149  {'media': [{'display_url': 'pic.twitter.com/BL...   
1991-02-08 16:08:05.002620928  {'media': [{'display_url': 'pic.twitter.com/r7...   

                               favorite_count  favorited  geo  \
1991-02-08 13:48:08.022790149            2530          0  NaN   
1991-02-08 16:08:05.002620928             126          0  NaN   

                                               id              id_str  \
1991-02-08 13:48:08.022790149  666020888022790144  666020888022790144   
1991-02-08 16:08:05.002620928  666029285002620928  666029285002620928   

                                                     ...                          \
1991-02-08 13:48:08.022790149                        ...                           
1991-02-08 16:08:05.002620928                        ...                           

                              quoted_status  quoted_status_id  \
1991-02-08 13:48:08.022790149           NaN               NaN   
1991-02-08 16:08:05.002620928           NaN               NaN   

                               quoted_status_id_str  retweet_count  retweeted  \
1991-02-08 13:48:08.022790149                   NaN            501          0   
1991-02-08 16:08:05.002620928                   NaN             47          0   

                               retweeted_status  \
1991-02-08 13:48:08.022790149               NaN   
1991-02-08 16:08:05.002620928               NaN   

                                                                          source  \
1991-02-08 13:48:08.022790149  <a href="http://twitter.com/download/iphone" r...   
1991-02-08 16:08:05.002620928  <a href="http://twitter.com/download/iphone" r...   

                                                                            text  \
1991-02-08 13:48:08.022790149  Here we have a Japanese Irish Setter. Lost eye...   
1991-02-08 16:08:05.002620928  This is a western brown Mitsubishi terrier. Up...   

                               truncated  \
1991-02-08 13:48:08.022790149          0   
1991-02-08 16:08:05.002620928          0   

                                                                            user  
1991-02-08 13:48:08.022790149  {'contributors_enabled': False, 'created_at': ...  
1991-02-08 16:08:05.002620928  {'contributors_enabled': False, 'created_at': ...  

[2 rows x 30 columns]

In [12]:
images.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [14]:
df_tweet.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2340 entries, 1991-02-08 13:48:08.022790149 to 1998-04-12 22:37:23.555336193
Data columns (total 30 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2340 non-null datetime64[ns]
entities                         2340 non-null object
extended_entities                1822 non-null object
favorite_count                   2340 non-null int64
favorited                        2340 non-null int64
geo                              0 non-null float64
id                               2340 non-null int64
id_str                           2340 non-null int64
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null float64
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 non-null float64
is_quote_status                  23

In [15]:
images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [16]:
df[df.isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id timestamp source  \
0          NaN                    NaN                  NaN       NaN    NaN   
1          NaN                    NaN                  NaN       NaN    NaN   
2          NaN                    NaN                  NaN       NaN    NaN   
3          NaN                    NaN                  NaN       NaN    NaN   
4          NaN                    NaN                  NaN       NaN    NaN   
5          NaN                    NaN                  NaN       NaN    NaN   
6          NaN                    NaN                  NaN       NaN    NaN   
7          NaN                    NaN                  NaN       NaN    NaN   
8          NaN                    NaN                  NaN       NaN    NaN   
9          NaN                    NaN                  NaN       NaN    NaN   
10         NaN                    NaN                  NaN       NaN    NaN   
11         NaN                    NaN                  NaN       NaN    NaN   
12         NaN                    NaN                  NaN       NaN    NaN   
13         NaN                    NaN                  NaN       NaN    NaN   
14         NaN                    NaN                  NaN       NaN    NaN   
15         NaN                    NaN                  NaN       NaN    NaN   
16         NaN                    NaN                  NaN       NaN    NaN   
17         NaN                    NaN                  NaN       NaN    NaN   
18         NaN                    NaN                  NaN       NaN    NaN   
19         NaN                    NaN                  NaN       NaN    NaN   
20         NaN                    NaN                  NaN       NaN    NaN   
21         NaN                    NaN                  NaN       NaN    NaN   
22         NaN                    NaN                  NaN       NaN    NaN   
23         NaN                    NaN                  NaN       NaN    NaN   
24         NaN                    NaN                  NaN       NaN    NaN   
25         NaN                    NaN                  NaN       NaN    NaN   
26         NaN                    NaN                  NaN       NaN    NaN   
27         NaN                    NaN                  NaN       NaN    NaN   
28         NaN                    NaN                  NaN       NaN    NaN   
29         NaN                    NaN                  NaN       NaN    NaN   
...        ...                    ...                  ...       ...    ...   
2326       NaN                    NaN                  NaN       NaN    NaN   
2327       NaN                    NaN                  NaN       NaN    NaN   
2328       NaN                    NaN                  NaN       NaN    NaN   
2329       NaN                    NaN                  NaN       NaN    NaN   
2330       NaN                    NaN                  NaN       NaN    NaN   
2331       NaN                    NaN                  NaN       NaN    NaN   
2332       NaN                    NaN                  NaN       NaN    NaN   
2333       NaN                    NaN                  NaN       NaN    NaN   
2334       NaN                    NaN                  NaN       NaN    NaN   
2335       NaN                    NaN                  NaN       NaN    NaN   
2336       NaN                    NaN                  NaN       NaN    NaN   
2337       NaN                    NaN                  NaN       NaN    NaN   
2338       NaN                    NaN                  NaN       NaN    NaN   
2339       NaN                    NaN                  NaN       NaN    NaN   
2340       NaN                    NaN                  NaN       NaN    NaN   
2341       NaN                    NaN                  NaN       NaN    NaN   
2342       NaN                    NaN                  NaN       NaN    NaN   
2343       NaN                    NaN                  NaN       NaN    NaN   
2344       NaN                    NaN                  N

In [17]:
df.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1621  684926975086034944                    NaN                  NaN   
888   759846353224826880                    NaN                  NaN   
1231  713177543487135744                    NaN                  NaN   
2196  668852170888998912                    NaN                  NaN   
1076  739606147276148736                    NaN                  NaN   

                      timestamp  \
1621  2016-01-07 02:38:10 +0000   
888   2016-07-31 20:21:02 +0000   
1231  2016-03-25 01:35:51 +0000   
2196  2015-11-23 18:02:38 +0000   
1076  2016-06-05 23:53:41 +0000   

                                                 source  \
1621  <a href="http://twitter.com/download/iphone" r...   
888   <a href="http://twitter.com/download/iphone" r...   
1231  <a href="http://twitter.com/download/iphone" r...   
2196  <a href="http://twitter.com/download/iphone" r...   
1076  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1621  Meet Bruiser &amp; Charlie. They are the best ...                  NaN   
888   This is Kirby. He's a Beneblip Cumberpat. Pret...                  NaN   
1231  This is Lance. Lance doesn't give a shit. 10/1...                  NaN   
2196  Say hello to Bobb. Bobb is a Golden High Fescu...                  NaN   
1076  Meet Benji. He just turned 1. Has already give...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1621                       NaN                        NaN   
888                        NaN                        NaN   
1231                       NaN                        NaN   
2196                       NaN                        NaN   
1076                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1621  https://twitter.com/dog_rates/status/684926975...                11   
888   https://twitter.com/dog_rates/status/759846353...                11   
1231  https://twitter.com/dog_rates/status/713177543...                10   
2196  https://twitter.com/dog_rates/status/668852170...                11   
1076  https://twitter.com/dog_rates/status/739606147...                 9   

      rating_denominator     name doggo floofer  pupper puppo  
1621                  10  Bruiser  None    None    None  None  
888                   10    Kirby  None    None    None  None  
1231                  10    Lance  None    None    None  None  
2196                  10     Bobb  None    None    None  None  
1076                  10    Benji  None    None  pupper  None

In [18]:
df['name'].value_counts()

None           745
a               55
Charlie         12
Oliver          11
Cooper          11
Lucy            11
Tucker          10
Penny           10
Lola            10
Bo               9
Winston          9
Sadie            8
the              8
Buddy            7
Daisy            7
Bailey           7
Toby             7
an               7
Scout            6
Leo              6
Stanley          6
Dave             6
Jax              6
Milo             6
Bella            6
Koda             6
Jack             6
Rusty            6
Oscar            6
George           5
              ... 
Stephanus        1
Molly            1
Coleman          1
Josep            1
Halo             1
Ronnie           1
Sonny            1
Koko             1
Einstein         1
Dante            1
Cuddles          1
Fabio            1
Godi             1
Tom              1
Herb             1
Lassie           1
Randall          1
Carper           1
Clybe            1
Shakespeare      1
Rontu            1
Bobby       

In [19]:
df_tweet.sample(5)

contributors  coordinates          created_at  \
1996-08-18 12:18:01.858686976           NaN          NaN 2017-03-11 01:15:58   
1993-02-24 17:03:03.004487680           NaN          NaN 2016-05-12 01:40:42   
1991-04-25 14:54:31.085670400           NaN          NaN 2015-12-04 01:40:29   
1996-02-07 13:16:42.998870016           NaN          NaN 2017-01-24 01:08:40   
1992-03-08 04:34:44.593901568           NaN          NaN 2016-02-17 18:49:22   

                                                                        entities  \
1996-08-18 12:18:01.858686976  {'hashtags': [], 'media': [{'display_url': 'pi...   
1993-02-24 17:03:03.004487680  {'hashtags': [], 'media': [{'display_url': 'pi...   
1991-04-25 14:54:31.085670400  {'hashtags': [], 'media': [{'display_url': 'pi...   
1996-02-07 13:16:42.998870016  {'hashtags': [], 'symbols': [], 'urls': [{'dis...   
1992-03-08 04:34:44.593901568  {'hashtags': [], 'media': [{'display_url': 'pi...   

                                                               extended_entities  \
1996-08-18 12:18:01.858686976  {'media': [{'display_url': 'pic.twitter.com/76...   
1993-02-24 17:03:03.004487680  {'media': [{'display_url': 'pic.twitter.com/Cn...   
1991-04-25 14:54:31.085670400  {'media': [{'display_url': 'pic.twitter.com/Pa...   
1996-02-07 13:16:42.998870016                                                NaN   
1992-03-08 04:34:44.593901568  {'media': [{'display_url': 'pic.twitter.com/8N...   

                               favorite_count  favorited  geo  \
1996-08-18 12:18:01.858686976           17187          0  NaN   
1993-02-24 17:03:03.004487680            5131          0  NaN   
1991-04-25 14:54:31.085670400             742          0  NaN   
1996-02-07 13:16:42.998870016           13326          0  NaN   
1992-03-08 04:34:44.593901568            2163          0  NaN   

                                               id              id_str  \
1996-08-18 12:18:01.858686976  840370681858686976  840370681858686976   
1993-02-24 17:03:03.004487680  730573383004487680  730573383004487680   
1991-04-25 14:54:31.085670400  672591271085670400  672591271085670400   
1996-02-07 13:16:42.998870016  823699002998870016  823699002998870016   
1992-03-08 04:34:44.593901568  700029284593901568  700029284593901568   

                                                     ...                          \
1996-08-18 12:18:01.858686976                        ...                           
1993-02-24 17:03:03.004487680                        ...                           
1991-04-25 14:54:31.085670400                        ...                           
1996-02-07 13:16:42.998870016                        ...                           
1992-03-08 04:34:44.593901568                        ...                           

                              quoted_status  quoted_status_id  \
1996-08-18 12:18:01.858686976           NaN               NaN   
1993-02-24 17:03:03.004487680           NaN               NaN   
1991-04-25 14:54:31.085670400           NaN               NaN   
1996-02-07 13:16:42.998870016           NaN               NaN   
1992-03-08 04:34:44.593901568           NaN               NaN   

                               quoted_status_id_str  retweet_count  retweeted  \
1996-08-18 12:18:01.858686976                   NaN           4854          0   
1993-02-24 17:03:03.004487680                   NaN           2290          0   
1991-04-25 14:54:31.085670400                   NaN            189          0   
1996-02-07 13:16:42.998870016                   NaN           2625          0   
1992-03-08 04:34:44.593901568                   NaN            622          0   

                               retweeted_status  \
1996-08-18 12:18:01.858686976               NaN   
1993-02-24 17:03:03.004487680               NaN   
1991-04-25 14:54:31.085670400               NaN   
1996-02-07 13:16:42.998870016               NaN   
1992-03-08 04:34:44.593901568               NaN   

                    

In [20]:
df.shape[0] - df[df['retweeted_status_id'].isnull()].shape[0]

181

In [21]:
images.sample(5)

tweet_id                                          jpg_url  \
100   667782464991965184  https://pbs.twimg.com/media/CURwm3cUkAARcO6.jpg   
959   705591895322394625  https://pbs.twimg.com/media/CcrEFQdUcAA7CJf.jpg   
943   704347321748819968  https://pbs.twimg.com/media/CcZYJniXEAAEJRF.jpg   
1262  748977405889503236  https://pbs.twimg.com/media/CmTm-XQXEAAEyN6.jpg   
309   671538301157904385  https://pbs.twimg.com/media/CVHIhi2WsAEgdKk.jpg   

      img_num                           p1   p1_conf  p1_dog  \
100         1                     lorikeet  0.466149   False   
959         1                      basenji  0.877207    True   
943         1                        teddy  0.233378   False   
1262        1  German_short-haired_pointer  0.742216    True   
309         1                   park_bench  0.194211   False   

                     p2   p2_conf  p2_dog                  p3   p3_conf  \
100         hummingbird  0.083011   False        African_grey  0.054247   
959   Italian_greyhound  0.047854    True  miniature_pinscher  0.035638   
943         feather_boa  0.088474   False    Brittany_spaniel  0.082917   
1262           bluetick  0.152810    True      English_setter  0.051835   
309        water_bottle  0.071870   False              beacon  0.053433   

      p3_dog  
100    False  
959     True  
943     True  
1262    True  
309    False

In [22]:
df_tweet.shape[0] - df_tweet[df_tweet['retweeted_status'].isnull()].shape[0]

167

### Quality

`df table:`
- some of the dog names contain normal words like 'actually', 'this', 'infuriating', 'getting', and so on. I will replace them with none.
- remove the retweets that are approximately 181 rows
- remove any columns that are not useful to the project

`Image predictions table:`
- Consistency issues with 3 columns
- p1: capitalize first letter
- p2: capitalize first letter
- p3: capitalize first letter
- p1: replace '_' with space
- p2: replace '_' with space
- p3: replace '_' with space

`df_tweet table:`
- remove the 176 rows of retweeted status id
- rename columns that mismatch names of other tables for merging purposes later like 'id' column

### Tidiness

`df table:`
- Date time columns need to be converted to datetime objects
- Rating column needs to converted to floats
- drop any columns that are not useful to the project
- combine dog stages like "pupper", "doggo" and the like into one column

`tweet_df table:`
- convert id column to string since its ids
- use the tweet_id as the index

- merge this table using join with the rest

### Clean `df_clean` table first

In [23]:
# create a separate copy of all 3 dataframes
df_clean = df.copy()
df_tweet_clean = df_tweet.copy()
images_clean = images.copy()

#### Define
- some of the dog names contain normal words like 'actually', 'this', 'infuriating', 'getting', and so on. I will replace them with none.

#### Code

In [24]:
# start off by replacing invalid names
# used excel to scan them visually and will replace all the faulty ones
df_clean['name'].replace(['actually', 'getting', 'infuriating', 'the', 'light', 'an', 'life',
                         'this', 'very', 'not', 'one', 'O', 'quite', 'such', 'old', 'all', 'by', 'a'], 
                         ['None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None',
                         'None', 'None', 'None', 'None', 'None', 'None', 'None'], inplace=True)

#### Test

In [25]:
df_clean['name'].sample(50) # ran the sample code multiple times and did not find any wrong name

95         Walter
820            Al
19         Canela
1653         None
407       Winston
2263         Kial
761       Stanley
267         Penny
20         Gerald
1968      Frankie
1727     Penelope
1082         None
101          None
2182     Harrison
1501         None
872          None
329         Poppy
2118        Damon
1375        Bilbo
601          Lola
414        Hobbes
490          Gary
1294       Farfle
1855        Ellie
194        George
142      Meatball
131          None
1917        Kenny
192          None
435       Peaches
1173        Derek
2145         None
1013      Atticus
1932         None
827        DonDon
2047    Scruffers
1595         None
742          None
2067         None
2045          Jax
2282    Genevieve
648         Robin
2295         Gabe
725         Penny
1865         None
295       Vincent
2117         Herb
143       Paisley
1250     Chuckles
332          None
Name: name, dtype: object

#### Define
- remove the retweets that are approximately 181 rows

#### Code

In [26]:
# remove rewtweeted id's that are not null
ret_index = df_clean[df_clean['retweeted_status_id'].isnull()==False].index
df_clean.drop(ret_index, inplace=True)

#### Test

In [27]:
df_clean['retweeted_status_id'].isna().sum()

2175

#### Define
- Combine dog stages like "pupper", "doggo" and the like into one column

#### Code

In [28]:
# combine four into one
df_clean['stage'] = df[['doggo', 'floofer','pupper','puppo']].apply(lambda x: ''.join(x), axis=1)
#replace combined values with valied ones
df_clean['stage'].replace(['NoneNoneNoneNone', 'doggoNoneNoneNone', 'NoneflooferNoneNone', 'NoneNonepupperNone', 'NoneNoneNonepuppo'],
                          ['None', 'doggo', 'floofer', 'pupper', 'puppo'], inplace=True)

In [29]:
# drop used columns
df_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], inplace=True, axis=1)

#### Test

In [30]:
df_clean.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name stage  
0                  10  Phineas  None  
1                  10    Tilly  None  
2                  10   Archie  None

#### Define
- remove any other columns that are not useful to the project

#### Code

In [31]:
df_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'stage'],
      dtype='object')

In [32]:
df_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'source', 'text', 'retweeted_status_id',
              'retweeted_status_user_id','retweeted_status_timestamp', 'expanded_urls'], axis=1, inplace=True)

#### Test

In [33]:
df_clean.head()

tweet_id                  timestamp  rating_numerator  \
0  892420643555336193  2017-08-01 16:23:56 +0000                13   
1  892177421306343426  2017-08-01 00:17:27 +0000                13   
2  891815181378084864  2017-07-31 00:18:03 +0000                12   
3  891689557279858688  2017-07-30 15:58:51 +0000                13   
4  891327558926688256  2017-07-29 16:00:24 +0000                12   

   rating_denominator      name stage  
0                  10   Phineas  None  
1                  10     Tilly  None  
2                  10    Archie  None  
3                  10     Darla  None  
4                  10  Franklin  None

#### Define
- convert timestamp column to datetime and extract date and time as separate columns

In [34]:
# convert to datetime
df_clean['timestamp'] = pd.to_datetime(df_clean['timestamp'])

In [35]:
# extract date and time
# got stuck with this for a while so googled how to
df_clean['date'] = df_clean['timestamp'].apply(lambda time: time.strftime('%m-%d-%Y'))
df_clean['time'] = df_clean['timestamp'].apply(lambda time: time.strftime('%H:%M:%S'))

In [36]:
# convert date and time column to datetime
df_clean['date'] = pd.to_datetime(df_clean['date'])
df_clean['time'] = pd.to_datetime(df_clean['time'])

In [37]:
# drop used column
df_clean.drop('timestamp', axis=1, inplace=True)

#### Test

In [38]:
df_clean.head()

tweet_id  rating_numerator  rating_denominator      name stage  \
0  892420643555336193                13                  10   Phineas  None   
1  892177421306343426                13                  10     Tilly  None   
2  891815181378084864                12                  10    Archie  None   
3  891689557279858688                13                  10     Darla  None   
4  891327558926688256                12                  10  Franklin  None   

        date                time  
0 2017-08-01 2019-01-21 16:23:56  
1 2017-08-01 2019-01-21 00:17:27  
2 2017-07-31 2019-01-21 00:18:03  
3 2017-07-30 2019-01-21 15:58:51  
4 2017-07-29 2019-01-21 16:00:24

#### Define
- tweet_id should be string even though the values are numbers

#### Code

In [39]:
df_clean['tweet_id'] = df_clean['tweet_id'].astype(str)

#### Test

In [40]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 7 columns):
tweet_id              2175 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
stage                 2175 non-null object
date                  2175 non-null datetime64[ns]
time                  2175 non-null datetime64[ns]
dtypes: datetime64[ns](2), int64(2), object(3)
memory usage: 135.9+ KB


## Clean `image_clean`

In [41]:
images_clean.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

#### Define
- remove underscores then capitalize first letters of p1, p2, and p3

In [42]:
# remove underscores
images_clean['p1'] = images_clean['p1'].str.replace('_', ' ')
images_clean['p2'] = images_clean['p2'].str.replace('_', ' ')
images_clean['p3'] = images_clean['p3'].str.replace('_', ' ')

# capitalize first letters
images_clean['p1'] = images_clean['p1'].str.title()
images_clean['p2'] = images_clean['p2'].str.title()
images_clean['p3'] = images_clean['p3'].str.title()

#### Test

In [43]:
images_clean.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh Springer Spaniel  0.465074    True              Collie   
1        1                 Redbone  0.506826    True  Miniature Pinscher   
2        1         German Shepherd  0.596461    True            Malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland Sheepdog  0.061428    True  
1  0.074192    True  Rhodesian Ridgeback  0.072010    True  
2  0.138584    True           Bloodhound  0.116197    True

#### Define
- convert tweet_id to string

#### Code

In [44]:
images_clean['tweet_id'] = images_clean['tweet_id'].astype(str)

#### Test

In [45]:
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


### Clean `df_tweet_clean`

#### Define
- rename 'id' column to 'tweet_id' to make joining possible later

#### Code

In [46]:
df_tweet_clean.rename(columns={'id' : 'tweet_id'}, inplace=True)

#### Test

In [47]:
df_tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2340 entries, 1991-02-08 13:48:08.022790149 to 1998-04-12 22:37:23.555336193
Data columns (total 30 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2340 non-null datetime64[ns]
entities                         2340 non-null object
extended_entities                1822 non-null object
favorite_count                   2340 non-null int64
favorited                        2340 non-null int64
geo                              0 non-null float64
tweet_id                         2340 non-null int64
id_str                           2340 non-null int64
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null float64
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 non-null float64
is_quote_status                  23

#### Define
- remove retweeted_status to get rid of non original tweets

In [49]:
df_tweet_ret_index = df_tweet_clean[df_tweet_clean['retweeted_status'].isnull()==False].index
df_tweet_clean.drop(df_tweet_ret_index, inplace=True)

#### Test

In [50]:
df_tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2173 entries, 1991-02-08 13:48:08.022790149 to 1998-04-12 22:37:23.555336193
Data columns (total 30 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2173 non-null datetime64[ns]
entities                         2173 non-null object
extended_entities                1747 non-null object
favorite_count                   2173 non-null int64
favorited                        2173 non-null int64
geo                              0 non-null float64
tweet_id                         2173 non-null int64
id_str                           2173 non-null int64
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null float64
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 non-null float64
is_quote_status                  21

#### Define
- Remove columns not needed for the project

#### Code

In [51]:
df_tweet_clean.columns

Index(['contributors', 'coordinates', 'created_at', 'entities',
       'extended_entities', 'favorite_count', 'favorited', 'geo', 'tweet_id',
       'id_str', 'in_reply_to_screen_name', 'in_reply_to_status_id',
       'in_reply_to_status_id_str', 'in_reply_to_user_id',
       'in_reply_to_user_id_str', 'is_quote_status', 'lang', 'place',
       'possibly_sensitive', 'possibly_sensitive_appealable', 'quoted_status',
       'quoted_status_id', 'quoted_status_id_str', 'retweet_count',
       'retweeted', 'retweeted_status', 'source', 'text', 'truncated', 'user'],
      dtype='object')

In [52]:
df_tweet_clean.drop(['contributors', 'coordinates', 'created_at', 'entities', 'extended_entities',
                    'favorited', 'geo', 'id_str', 'in_reply_to_screen_name', 'in_reply_to_status_id', 
                    'in_reply_to_status_id_str', 'in_reply_to_user_id', 
                    'in_reply_to_user_id_str','is_quote_status', 'lang', 'place', 
                    'possibly_sensitive', 'possibly_sensitive_appealable',
                    'quoted_status', 'quoted_status_id', 'quoted_status_id_str', 'retweeted', 'retweeted_status',
                    'source', 'truncated', 'user'], axis=1, inplace=True)

#### Test

In [53]:
df_tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2173 entries, 1991-02-08 13:48:08.022790149 to 1998-04-12 22:37:23.555336193
Data columns (total 4 columns):
favorite_count    2173 non-null int64
tweet_id          2173 non-null int64
retweet_count     2173 non-null int64
text              2173 non-null object
dtypes: int64(3), object(1)
memory usage: 84.9+ KB


#### Define
- tweet_id column needs to be string

#### Code

In [54]:
df_tweet_clean['tweet_id'] = df_tweet_clean['tweet_id'].astype(str)

#### Test

In [55]:
df_tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2173 entries, 1991-02-08 13:48:08.022790149 to 1998-04-12 22:37:23.555336193
Data columns (total 4 columns):
favorite_count    2173 non-null int64
tweet_id          2173 non-null object
retweet_count     2173 non-null int64
text              2173 non-null object
dtypes: int64(2), object(2)
memory usage: 84.9+ KB
